clang/lib/Driver/Driver.cpp

```diff
@@ -898,6 +898,29 @@ void Driver::CreateOffloadingDeviceToolChains(Compilation &C,
                                                << "-ffreestanding";
   }
 
+  // Diagnose incorrect inputs to SYCL options.
+  // FIXME: Since the option definition includes the list of possible values,
+  // the validation must be automatic, not requiring separate disjointed code
+  // blocks accross the driver code. Long-term, the detection of incorrect
+  // values must happen at the level of TableGen and Arg class design, with
+  // Compilation/Driver class constructors handling the driver-specific
+  // diagnostic output.
+  auto checkSingleArgValidity = [&](Arg *A,
+                                    SmallVector<StringRef, 4> AllowedValues) {
+    if (!A)
+      return;
+    const char *ArgValue = A->getValue();
+    for (const StringRef AllowedValue : AllowedValues)
+      if (AllowedValue.equals(ArgValue))
+        return;
+    Diag(clang::diag::err_drv_invalid_value)
+        << A->getOption().getName() << ArgValue;
+  };
+  checkSingleArgValidity(SYCLLink, {"early", "image"});
+  checkSingleArgValidity(
+      C.getInputArgs().getLastArg(options::OPT_fsycl_device_code_split_EQ),
+      {"per_kernel", "per_source", "auto", "off"});
```

Comment:
Another option that can be checked is `-fsycl-help`.  I don't see any other SYCL based options that require a specific set of arguments.

Reply:
`-fsycl-help` input validity is currently being checked within `Driver::PrintSYCLToolHelp()` - the same unsupported argument diagnostic gets emitted there. I could **a)** still rework it so that validity checks are done at one site instead of 2+ (which is kind of the point of the change), or **b)** leave the current code to minimize duplication. My personal slight preference is towards **b)**, but I'm happy to adjust this based on your advice.

Reply:
I'm OK to leave as is as -fsycl-help is performing it's own check.